### PR TITLE
feat: sort, filters and search products

### PR DESCRIPTION
**Pull request by ravikumar0403**
Change log:
- sort by price
- sort by rating
- filter by category
- filter by price
- filter by rating

### CODE REVIEW

File: src/components/Filters.js [hunk 1]
```diff
@@ -1,34 +1,89 @@
...
+          }
+        >
+          Clear filters
+        </button>
       </div>
       <div className="my-1">
         <p className="fs-2">Sort by</p>
-        <select className="input">
-          <option value="" defaultValue>
+        <select className="input" onChange={handleSortChange}>
```
Comment (yashpatel1797) [inline src/components/Filters.js:55]: I like how you have created a handler so that HTML looks clean.

File: src/utils/price.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/**
+ * Filter json array based on rating prop
+ * @param {array} products - Array of product objects
+ * @param {number} price - minimum price.
+ * @returns {array} - Filtered array of objects.
+ */
```
Comment (yashpatel1797) [inline src/utils/price.js:6]: nice that you have added details also.

File: src/context/StateProvider.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { FiltersProvider } from "./filter-context";
+import { ProductProvider } from "./product-context";
+
+const StateProvider = ({ children }) => {
+  return (
+    <ProductProvider>
```
Comment (bharati-21) [inline src/context/StateProvider.js:6]: Great Idea to wrap necessary providers together and then return them together!

File: src/context/filter-context.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import { createContext, useContext, useReducer } from "react";
+import { filterReducer } from "../reducers";
+
+const filterContext = createContext();
```
Comment (bharati-21) [inline src/context/filter-context.js:4]: Convention followed is that the created context object is written in PascalCase. Maybe you could adopt it.

File: src/utils/category.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Filter json array based on category prop
+ * @param {array} products - Array of product objects
```
Comment (bharati-21) [inline src/utils/category.js:3]: Writing a gist of what the file does and returns is a great way of providing a summary.

**Review by yashpatel1797 — APPROVED** [review]
All the features added are working as expected. and naming and code structure is also well written.

File: src/utils/sortObj.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Sorts json array.
+ * @param {array} products - Array of product objects.
+ * @param {string} sortBy - sort condition.
+ * @returns {array} - sorted array of objects.
+ */
+export const getSortedProducts = (products, sortBy) => {
+  switch (sortBy) {
+    case "price_asc":
+      return [...products.sort((a, b) => a.price - b.price)];
```
Comment (bharati-21) [inline src/utils/sortObj.js:10]: Nice touch by using pure functions!

**Review by bharati-21 — APPROVED** [review]
Very well written and the code is very easy to understand. Good naming conventions and great file structures.

2 things I noticed -
- Sidebar filter menu overflow. 
image
- When a user enters anything on the search bar, any filters are applied to the products and then cleared, the initial list itself is returned. Maybe you could then remove the search text when this happens or try getting back the products that were filtered based on the search text? 


Apart from this all looks good to me.

**Comment by ravikumar0403** [comment]
> Very well written and the code is very easy to understand. Good naming conventions and great file structures.
> 
> 2 things I noticed -
> 
> * Sidebar filter menu overflow.
>   image
> * When a user enters anything on the search bar, any filters are applied to the products and then cleared, the initial list itself is returned. Maybe you could then remove the search text when this happens or try getting back the products that were filtered based on the search text?
> 
> Apart from this all looks good to me.

Thanks for bringing this to my notice. I'll fix these issues.

File: src/context/filter-context.js (hunk 4 above)
Comment (ravikumar0403) [reply src/context/filter-context.js:4]: Sure, will make this change

**Pull request merged**